File: app/src/main/java/com/example/david/lists/view/userlistlist/UserListListView.java
package com.example.david.lists.view.userlistlist;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.example.david.lists.R;
import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.util.UtilExceptions;
import com.example.david.lists.util.UtilNightMode;
import com.example.david.lists.view.addedit.userlist.AddEditUserListDialog;
import com.example.david.lists.view.authentication.AuthView;
import com.example.david.lists.view.authentication.IAuthContract;
import com.example.david.lists.view.common.ListViewBase;
import com.example.david.lists.view.itemlist.ItemActivity;
import com.example.david.lists.view.userlistlist.buldlogic.DaggerUserListListViewComponent;

import javax.inject.Inject;

public class UserListListView extends ListViewBase
        implements IUserListViewContract.View,
        ConfirmSignOutDialog.ConfirmSignOutCallback {

    @Inject
    IUserListViewContract.Logic logic;

    public UserListListView() {
    }

    public static UserListListView newInstance() {
        return new UserListListView();
    }


    @Override
    public void onAttach(Context context) {
        inject();
        super.onAttach(context);
    }

    private void inject() {
        DaggerUserListListViewComponent.builder()
                .application(getActivity().getApplication())
                .view(this)
                .movementCallback(this)
                .build()
                .inject(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        logic.onStart();
        return view;
    }

    @Override
    public void onDestroy() {
        logic.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(logic.getMenuResource(), menu);
        menu.findItem(R.id.menu_id_night_mode)
                .setChecked(UtilNightMode.isNightModeEnabled(getActivity().getApplication()));
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_id_sign_out:
                logic.signOut();
                break;
            case R.id.menu_id_sign_in:
                logic.signIn();
                break;
            case R.id.menu_id_night_mode:
                logic.nightMode(item);
                break;
            default:
                UtilExceptions.throwException(new IllegalArgumentException());
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    public void openUserList(UserList userList) {
        Intent intent = new Intent(getActivity(), ItemActivity.class);
        intent.putExtra(getString(R.string.intent_extra_user_list_id), userList.getId());
        intent.putExtra(getString(R.string.intent_extra_user_list_title), userList.getTitle());
        startActivity(intent);
    }


    @Override
    public void confirmSignOut() {
        openDialogFragment(new ConfirmSignOutDialog());
    }

    @Override
    public void signOutConfirmed() {
        logic.signOutConfirmed();
    }

    @Override
    public void openAuthentication(IAuthContract.AuthGoal authGoal, int requestCode) {
        Intent intent = new Intent(getActivity(), AuthView.class);
        intent.putExtra(getString(R.string.intent_extra_auth), authGoal);
        startActivityForResult(intent, requestCode);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        logic.authResult(requestCode, data);
    }


    @Override
    public void openAddDialog(int position) {
        openDialogFragment(AddEditUserListDialog.getInstance(
                "", "", position
        ));
    }

    @Override
    public void openEditDialog(UserList userList) {
        openDialogFragment(AddEditUserListDialog.getInstance(
                userList.getId(), userList.getTitle(), userList.getPosition()
        ));
    }


    @Override
    public void notifyUserOfDeletion(String message) {
        notifyDeletionSnackbar(message);
    }


    @Override
    public void setStateDisplayList() {
        displayList();
    }

    @Override
    public void setStateLoading() {
        displayLoading();
    }

    @Override
    public void setStateError(String message) {
        displayError(message);
    }

    @Override
    public void recreateView() {
        getActivity().recreate();
    }


    @Override
    protected void addButtonClicked() {
        logic.add();
    }

    @Override
    protected void undoRecentDeletion() {
        logic.undoRecentDeletion();
    }

    @Override
    protected void deletionNotificationTimedOut() {
        logic.deletionNotificationTimedOut();
    }

    @Override
    protected void draggingListItem(int fromPosition, int toPosition) {
        logic.dragging(fromPosition, toPosition);
    }

    @Override
    protected void permanentlyMoved(int newPosition) {
        logic.movedPermanently(newPosition);
    }

    @Override
    protected String getTitle() {
        return getString(R.string.app_name);
    }

    @Override
    protected boolean enableUpNavigationOnToolbar() {
        return false;
    }

    @Override
    protected RecyclerView.Adapter getAdapter() {
        return logic.getAdapter();
    }
}

File: app/src/main/java/com/example/david/lists/view/userlistlist/UserListListView.kt
package com.example.david.lists.view.userlistlist

import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.example.david.lists.R
import com.example.david.lists.data.datamodel.UserList
import com.example.david.lists.util.UtilExceptions
import com.example.david.lists.view.common.ListViewBase
import com.example.david.lists.view.userlistlist.buldlogic.DaggerUserListListViewComponent
import javax.inject.Inject

class UserListListView : ListViewBase(),
        IUserListViewContract.View {

    @Inject
    lateinit var logic: IUserListViewContract.Logic

    @Inject
    lateinit var adapter: IUserListViewContract.Adapter

    override val title: String
        get() = getString(R.string.app_name)


    override fun onAttach(context: Context) {
        inject()
        super.onAttach(context)
    }

    private fun inject() {
        DaggerUserListListViewComponent.builder()
                .application(activity!!.application)
                .view(this)
                .fragment(this)
                .movementCallback(this)
                .build()
                .inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        logic.onStart()
    }

    override fun onDestroy() {
        logic.onDestroy()
        super.onDestroy()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_sign_out, menu)
        initMenuSetCheckedState(menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    private fun initMenuSetCheckedState(menu: Menu) {
        menu.findItem(R.id.menu_id_night_mode).isChecked = logic.isNightModeEnabled
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_id_sign_out -> logic.signOut()
            R.id.menu_id_night_mode -> {
                with(item.isChecked) {
                    logic.setNightMode(this)
                    this.not()
                }
            }
            else -> UtilExceptions.throwException(IllegalArgumentException())
        }
        return super.onOptionsItemSelected(item)
    }


    override fun openUserList(userList: UserList) {
        navController.get().navigate(UserListListViewDirections.actionUserListListViewToItemListView(
                userList.id,
                userList.title
        ))
    }


    override fun confirmSignOut() {
        navController.get().navigate(
                UserListListViewDirections.actionUserListListViewToConfirmSignOutDialog()
        )
    }


    override fun openAddDialog(position: Int) {
        navController.get().navigate(
                UserListListViewDirections.actionUserListListViewToAddEditUserListDialog(
                        "", "", position
                )
        )
    }

    override fun openEditDialog(userList: UserList) {
        navController.get().navigate(
                UserListListViewDirections.actionUserListListViewToAddEditUserListDialog(
                        userList.id, userList.title, userList.position
                )
        )
    }

    override fun setViewData(viewData: List<UserList>) {
        adapter.setData(viewData)
    }

    override fun notifyUserOfDeletion(message: String) {
        notifyDeletionSnackbar(message)
    }


    override fun setStateDisplayList() {
        displayList()
    }

    override fun setStateLoading() {
        displayLoading()
    }

    override fun setStateError(message: String) {
        displayError(message)
    }


    override fun addButtonClicked() {
        logic.add()
    }

    override fun undoRecentDeletion() {
        logic.undoRecentDeletion(adapter)
    }

    override fun deletionNotificationTimedOut() {
        logic.deletionNotificationTimedOut()
    }

    override fun draggingListItem(fromPosition: Int, toPosition: Int) {
        logic.dragging(fromPosition, toPosition, adapter)
    }

    override fun permanentlyMoved(newPosition: Int) {
        logic.movedPermanently(newPosition)
    }

    override fun enableUpNavigationOnToolbar() = false

    override fun getAdapter() = adapter as RecyclerView.Adapter<*>
}
